k8s: Introduce subscriber package to simplify & consolidate K8s watcher callbacks / event handling

christarazi opened the pull request:
This PR implements a new package which represents a subscriber to K8s watcher
events. This package simplifies the logic of K8s watchers and removes all the
unnecessary logic of the consumer (subscriber) from them.

In order for a subscriber to add their event handling to the K8s watcher, it
must use register itself. Depending on what kind of object it's subscribed to,
it must use implement the ServiceHandler interface if it's handling Service
objects, or use the "Raw" API where no such constraint on the object type
exists.

With this implementation, future subscribers can easily add their own logic
without worrying about polluting K8s watcher code with specific, irrelevant
details only really meant for them. Meanwhile, the K8s watchers themselves
simply call `Notify*()` depending on the event type. This notifies all the
subscribers to run their logic on the event.

In an upcoming PR, a BGP subscriber will utilize this infrastructure, all the
while, the core service watcher code remains unchanged.

This PR does not convert the rest of the Operator's watchers, but instead
provides the groundwork for future PRs.

See commit msgs.

- watchers: Add subscriber package
- operator/watchers: Convert service watcher to use subscriber

aanm commented on operator/watchers/queue/queue.go:
```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Authors of Cilium
```
Can't we make this a library instead of defining everything globally in package queue?

aanm commented on operator/watchers/queue/queue_test.go:
```diff
@@ -0,0 +1,42 @@
+package queue
```
Missing copyright header.

aanm commented on operator/watchers/queue/queue.go:
```diff
@@ -0,0 +1,133 @@
+
+// Stopped returns whether the queue is stopped.
+func Stopped() bool {
+	return stopped.Load().(bool)
```
We can avoid using the atomic value if we have something like this:
```golang
	select {
	case <-stopCh:
		return true
	default:
		return false
	}
```

aanm commented on operator/watchers/queue/queue.go:
```diff
@@ -0,0 +1,133 @@
+func Stop() {
+	if !q.ShuttingDown() {
+		close(stopCh)
+	}
```
If the caller executes `Stop()` multiple times, line 86 will panic because the channel is already closed.

aanm reviewed (changes requested):
It's not entirely clear what are we gaining by introducing another queue to process k8s events. We had queues for k8s events in the past but we have removed them because they are redundant as the k8s watcher library already has a really good implementation to handle them.

christarazi committed:
watchers: Add subscriber package

This commit implements a package called subscriber for the K8s resource
/ object watchers. It contains a list of subscribers to notify when
handling a K8s add, update, or delete event, respectively.

This allows the watchers to be simpler and leaner than they previously
were. They no longer need to hold logic specific to the consumers.
Instead, they call Notify*() inside the watcher callbacks depending on
the event, passing the resource / object from the event. This notifies
all the subscribers registered with Register().

Subscribers (also referred to as consumers or handlers) must implement
either k8s.io/client-go/tools/cache.ResourceEventHandler and use the
"Raw" API of callbacks such as (*RawList).Register(), where there's no
strict contraint on the resource / object type, or implement the
provided ServiceHandler and use the "Service" specific API, such as
(*ServiceList).Register(). Consumers of the K8s Service watcher should
implement the ServiceHandler.

For now, only the raw and the service API is provided, due to strict
necessity. More specific handlers can be added easily by defining a new
interface similar to ServiceHandler, but differing on the object type.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
operator/watchers: Convert service watcher to use subscriber

This commit converts the service watcher in the Operator to use the new
subscriber package to handle K8s object events by defining a
ServiceHandler.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/watchers/callbacks/callbacks.go:
```diff
@@ -0,0 +1,59 @@
+	q.RLock()
+	defer q.RUnlock()
+	return q.cbs
+
```
Extra space

christarazi commented on pkg/k8s/watchers/callbacks/callbacks.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// Register registers an event handler's callbacks into the list.
+func (q *List) Register(cb cache.ResourceEventHandler) {
```
Nit: `q` leftover from previous implemention, should be `l`

christarazi commented on pkg/k8s/watchers/callbacks/callbacks.go:
```diff
@@ -0,0 +1,59 @@
+	q.Unlock()
+}
+
+// Callbacks returns all the event handler callbacks from the list.
```
Mention that the caller must not modify the slice

aanm commented on pkg/k8s/watchers/callbacks/callbacks.go:
```diff
@@ -0,0 +1,59 @@
+func (q *List) Callbacks() []cache.ResourceEventHandlerFuncs {
+	q.RLock()
+	defer q.RUnlock()
+	return q.cbs
```
The mutex will not protect the concurrent writes into the elements of `q.cbs` by the caller of this function.

twpayne replied:
Yes, this should probably return a copy of `q.cbs` in case `q.cbs` is modified after (e.g. by someone calling `q.Register(...)`) after `q.Callbacks()` returns.

aanm commented on pkg/k8s/watchers/callbacks/callbacks.go:
```diff
@@ -0,0 +1,59 @@
+		AddFunc:    cb.OnAdd,
+		UpdateFunc: cb.OnUpdate,
+		DeleteFunc: cb.OnDelete,
+	})
```
It is totally fine to define our own interfaces for each handler. I.e., have an interface that is specifically to handle Services, instead of `interfaces{}`, this will make allow us to not cast `interfaces{}` into the specific structures as we already did that in the watchers..

aanm commented on operator/watchers/service_cache.go:
```diff
@@ -0,0 +1,60 @@
+	svc, ok := obj.(*slim_corev1.Service)
+	if !ok {
+		return
+	}
```
I'm referring to this that can be avoided as explained in the other comment.

aditighag commented on operator/watchers/service_cache.go:
```diff
@@ -0,0 +1,48 @@
+
+func (c *serviceCacheSubscriber) OnAdd(obj *slim_corev1.Service) {
+	log.WithField(logfields.ServiceName, obj.Name).Debugf("Received service addition %+v", obj)
+	K8sSvcCache.UpdateService(obj, c.swgSvcs)
```
`K8sSvcCache.UpdateService` sends the events to a buffered channel of some pre-configured capacity. Currently, we have only one buffered channel as `K8sSvcCache.UpdateService` is directly called from the watcher callback. But with this change there'll be a channel for every subscriber. Won't that increase memory usage of Cilium? Also, it looks like the [code](https://github.com/cilium/cilium/blob/master/pkg/k8s/service_cache.go#L183..201) to parse the `Service` object, call `correlateEndpoints`, etc in `K8sSvcCache.UpdateService` be duplicated for every subscriber?

christarazi replied:
No, that's not the intent. I opened https://github.com/cilium/cilium/issues/15471 to further address subscribers that depend on the service cache itself. An example is explained there.

Generally , the solution wouldn't be to create a new channel or duplicate anything for that matter for each subscriber, but rather refactor the downstream subscribers that _depend_ on the service cache itself.

aanm reviewed (changes requested):
As this is part of the BGP PR, do we still want to keep this PR open?
